opte/src/engine/icmpv6.rs

```diff
@@ -172,3 +180,405 @@ impl HairpinAction for Icmpv6EchoReply {
+
+    Ok((target_addr, has_ll_option))
+}
+
+// Return the destination IP and a Neighbor Advertisement, based on the data
+// provided in a Neighbor Solicitation. If we should not generate an NA in
+// response to the NS, then `None` is returned.
+//
+// See https://www.rfc-editor.org/rfc/rfc4861.html#section-7.2.4 for details on
+// the validation and construction performed here.
+fn construct_neighbor_advert<'a>(
+    na: &'a NeighborAdvertisement,
+    target_addr: &'a Ipv6Addr,
+    has_ll_option: bool,
+    src_ip: &'a Ipv6Addr,
+) -> Option<(Ipv6Addr, NdiscRepr<'a>)> {
+    // Drop the packet if the target address is not actually our own address.
+    if target_addr != na.ip() {
+        return None;
+    }
+
+    // Set the ROUTER flag, if required.
+    let mut flags = NdiscNeighborFlags::empty();
+    flags.set(NdiscNeighborFlags::ROUTER, na.is_router);
+
+    // Even though this is NA is in response to an NS, if the source IP is
+    // UNSPEC, we must _not_ set the SOLICITED flag.
+    let src_is_unspec = src_ip == &Ipv6Addr::ANY_ADDR;
+    flags.set(NdiscNeighborFlags::SOLICITED, !src_is_unspec);
+
+    // Set the OVERRIDE flag.
+    //
+    // This should be clear if the target address is anycast or we're proxying,
+    // neither of which will ever be true. But it _should_ be set if the
+    // Link-Layer address option is included in the message, regardless of the
+    // addresses.
+    flags.set(NdiscNeighborFlags::OVERRIDE, has_ll_option);
+
+    // The destination IP address also depends on the source IP address.
+    //
+    // If the source is UNSPEC, we're required to send this to the all-nodes
+    // multicast group. Otherwise, we must unicast the NA back to the source.
+    let dst_ip = if src_is_unspec { Ipv6Addr::ALL_NODES } else { *src_ip };
+    Some((
+        dst_ip,
+        NdiscRepr::NeighborAdvert {
+            flags,
+            target_addr: Ipv6Address::from(*target_addr),
+            // Always include the Link-Layer address option.
+            lladdr: Some(RawHardwareAddress::from_bytes(&na.mac.bytes())),
+        },
+    ))
+}
+
+impl HairpinAction for NeighborAdvertisement {
+    fn implicit_preds(&self) -> (Vec<Predicate>, Vec<DataPredicate>) {
+        // The source IP must be a link-local IPv6 address, or, if
+        // `allow_unspec` is true, the unspecified address.
+        let source_addrs = if self.allow_unspec {
+            vec![
+                Ipv6AddrMatch::Prefix(Ipv6Cidr::LINK_LOCAL),
+                Ipv6AddrMatch::Exact(Ipv6Addr::ANY_ADDR),
+            ]
+        } else {
+            vec![Ipv6AddrMatch::Prefix(Ipv6Cidr::LINK_LOCAL)]
+        };
+
+        let hdr_preds = vec![
+            // We expect that the source MAC is the MAC provided to the client.
+            Predicate::InnerEtherSrc(vec![EtherAddrMatch::Exact(self.src_mac)]),
+            // It's directed to the virtual gateway's MAC address.
+            Predicate::InnerEtherDst(vec![EtherAddrMatch::Exact(self.mac)]),
```

Comment:
This is not correct. We're only allowing NSs unicast to the gateway's MAC, which is clearly not correct in some cases. This could be multicast to all-nodes, or to the multicast MAC address derived solicited-node multicast group. In the first case that would be `33:33:0:0:0:2`, and in the second it would be `33:33:ff:ff:77:77`.

IPv6 is _so_ complicated.